File: 0113-path-sum-ii/0113-path-sum-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    List<List<Integer>> res=new ArrayList<>();      // Global Variable dec; 2D Array Resultant wla

    public void fun(TreeNode root,int sum,List<Integer> diary,int targetSum){
        // Agar Tree ka Root hi null hai
        if(root==null){
            return;
        }

        sum+=root.val;                              // Sum mei Root k val k sath add kro
        diary.add(root.val);                        // Root ka value Diary mei push kro

        // Agar Leaf Node par aa gya
        if(root.left==null && root.right==null){
            if(sum==targetSum){
                res.add(new ArrayList<>(diary));    // Agar Target achieve hua toh Diary wla Res m daalo
            }
            diary.remove(diary.size()-1);           // Diary se hta k reverse m daalo
            return;
        }

        // Agar Non Leaf Node hai; extend from Leftwards and Rightwards
        fun(root.left,sum,diary,targetSum);         // Left wley se extend kro
        fun(root.right,sum,diary,targetSum);        // Right wley se extend kro

        diary.remove(diary.size()-1);               // Ye Non Leaf wlo k liye bhi toh pop back krna hoga

        return;
    }

    public List<List<Integer>> pathSum(TreeNode root,int targetSum) {

        List<Integer> diary=new ArrayList<>();

        // Recursive Function Call kardo
        fun(root,0,diary,targetSum);

        return res;
    }
}

File: 0113-path-sum-ii/0113-path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
vector<vector<int>>res;                                    // Global variable dec ; 2D array resultant wla
void fun(TreeNode*root,int sum,vector<int>&diary,int targetSum){
    // Agar tree ka root hi null hai 
    if(root==nullptr){
        return;
    }

    sum+=root->val;                                   // sum mei root k val k sath add kro 
    diary.push_back(root->val);                      // root ka value diary mei  push kro

    // agar leaf node par agya 
    if(root->left==nullptr && root->right==nullptr){
        if(sum==targetSum){
            res.push_back(diary);                     // agar target achieve hua toh diary wla res m dalo 
        }
        diary.pop_back();                         // diary se hta k reverse m dalo 
        return;
    }

    // Agar Non leaf node hai ; extend from leftwards and rightwards 
    fun(root->left,sum,diary,targetSum);                 // Left wley se extend kro 
    fun(root->right,sum,diary,targetSum);                // Right wley se extend kro 

    diary.pop_back();                                    // Y non leaf wlo k liye bhi toh pop back krna hoga 

    return;
}

    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        vector<int> diary;
        
        // Recursivee function call kardo 
        fun(root,0,diary,targetSum);

        return res;
    }
};